lib/spack/spack/config.py

```diff
@@ -206,61 +216,340 @@ def __repr__(self):
-#: No site-level configs should be checked into spack by default.
-_site_path = os.path.join(spack.etc_path, 'spack')
-ConfigScope('site', _site_path)
-ConfigScope('site/%s' % _platform, os.path.join(_site_path, _platform))
+        If ``scope`` is ``None`` or not provided, return the merged contents
+        of all of Spack's configuration scopes.  If ``scope`` is provided,
+        return only the confiugration as specified in that scope.
 
-#: User configuration can override both spack defaults and site config.
-_user_path = spack.user_config_path
-ConfigScope('user', _user_path)
-ConfigScope('user/%s' % _platform, os.path.join(_user_path, _platform))
+        This off the top-level name from the YAML section.  That is, for a
+        YAML config file that looks like this::
 
+           config:
+             install_tree: $spack/opt/spack
+             module_roots:
+               lmod:   $spack/share/spack/lmod
 
-def highest_precedence_scope():
-    """Get the scope with highest precedence (prefs will override others)."""
-    return list(config_scopes.values())[-1]
+        ``get_config('config')`` will return::
 
+           { 'install_tree': '$spack/opt/spack',
+             'module_roots: {
+                 'lmod': '$spack/share/spack/lmod'
+             }
+           }
 
-def validate_scope(scope):
-    """Ensure that scope is valid, and return a valid scope if it is None.
+        """
+        _validate_section_name(section)
+        merged_section = syaml.syaml_dict()
 
-       This should be used by routines in ``config.py`` to validate
-       scope name arguments, and to determine a default scope where no
-       scope is specified.
+        if scope is None:
+            scopes = self.scopes.values()
+        else:
+            scopes = [self._validate_scope(scope)]
+
+        for scope in scopes:
+            # read potentially cached data from the scope.
+
+            data = scope.get_section(section)
+
+            # Skip empty configs
+            if not data or not isinstance(data, dict):
+                continue
+
+            if section not in data:
+                continue
+
+            merged_section = _merge_yaml(merged_section, data)
+
+        # no config files -- empty config.
+        if section not in merged_section:
+            return {}
+
+        # take the top key off before returning.
+        return merged_section[section]
+
+    def get(self, path, default=None, scope=None):
+        """Get a config section or a single value from one.
+
+        Accepts a path syntax that allows us to grab nested config map
+        entries.  Getting the 'config' section would look like::
+
+            spack.config.get('config')
 
+        and the ``dirty`` section in the ``config`` scope would be::
+
+            spack.config.get('config:dirty')
+
+        We use ``:`` as the separator, like YAML objects.
+
+        TODO: Currently only handles maps. Think about lists if neded.
```

Comment:
Do you mean lists of objects? Something like:
```
spack.config.get('config:tests:0:url')
```
to get the `url` key from the item at index 0 of a list of things that looks like:
```python
[{ url: 'foo' },
 { url: 'bar'}]
```
? Or are you referring to another behavior that is not handled?

Reply:
I mean something like that, but I think we'd need to differentiate the syntax, e.g.:

```python
spack.config.get('config:tests[0]:url')
```

The keys following `:` are assumed to be strings, and result in `index[key]` in the Python code.  The indexes for lists are integers, and you'd need to convert them before indexing.  Yes, we could do that by looking at the type of structure we're indexing, but I think the above is clearer anyway because you cana clearly see what's expected to be a list and what's expected to be a dict in the path specifier.

I'm basically just saying we're not doing anything to address the list case currently, so you can't ask for, e.g.:

```python
spack.config.get('repos[2]')
```

Reply:
This function is used frequently so IMO the `TODO` should be a `#` comment vs. living in the docstring.